File: ruby/lib/natsuzora/renderer.rb
# frozen_string_literal: true

module Natsuzora
  class Renderer
    def initialize(ast, template_loader: nil)
      @ast = ast
      @template_loader = template_loader
      @escape_enabled = true
    end

    def render(data)
      @context = Context.new(data)
      render_nodes(@ast.nodes)
    end

    private

    def render_nodes(nodes)
      nodes.map { |node| render_node(node) }.join
    end

    def render_node(node)
      case node
      when AST::Text
        render_text(node)
      when AST::Variable
        render_variable(node)
      when AST::IfBlock
        render_if(node)
      when AST::UnlessBlock
        render_unless(node)
      when AST::EachBlock
        render_each(node)
      when AST::UnsecureBlock
        render_unsecure(node)
      when AST::Include
        render_include(node)
      else
        raise RenderError, "Unknown node type: #{node.class}"
      end
    end

    def render_text(node)
      node.content
    end

    def render_variable(node)
      value = @context.resolve(node.path)
      str = Value.stringify(value)
      @escape_enabled ? HtmlEscape.escape(str) : str
    end

    def render_if(node)
      value = @context.resolve(node.condition.path)
      if Value.truthy?(value)
        render_nodes(node.then_nodes)
      elsif node.else_nodes
        render_nodes(node.else_nodes)
      else
        ''
      end
    end

    def render_unless(node)
      value = @context.resolve(node.condition.path)
      if Value.truthy?(value)
        ''
      else
        render_nodes(node.body_nodes)
      end
    end

    def render_each(node)
      collection = @context.resolve(node.collection.path)
      Value.ensure_array!(collection)

      collection.each_with_index.map do |item, index|
        bindings = { node.item_name => item }
        bindings[node.index_name] = index if node.index_name

        @context.with_scope(bindings) do
          render_nodes(node.body_nodes)
        end
      end.join
    end

    def render_unsecure(node)
      prev_escape = @escape_enabled
      @escape_enabled = false
      result = render_nodes(node.nodes)
      @escape_enabled = prev_escape
      result
    end

    def render_include(node)
      raise IncludeError, 'Template loader not configured for include' unless @template_loader

      partial_ast = @template_loader.load(node.name)

      bindings = {}
      node.args.each do |key, var|
        bindings[key] = @context.resolve(var.path)
      end

      @template_loader.with_include(node.name) do
        @context.with_scope(bindings, include_scope: true) do
          render_nodes(partial_ast.nodes)
        end
      end
    end
  end
end

File: rust/crates/natsuzora/src/renderer.rs
//! Renderer for evaluating Natsuzora AST.
//!
//! Since TokenProcessor handles whitespace control before parsing,
//! the renderer simply evaluates the AST without any whitespace trimming logic.

use crate::context::Context;
use crate::error::{NatsuzoraError, Result};
use crate::html_escape;
use crate::template_loader::TemplateLoader;
use crate::value::Value;
use crate::ast::{
    AstNode, EachBlock, IfBlock, IncludeNode, Modifier, Template, UnlessBlock, UnsecureNode,
    VariableNode,
};
use std::collections::HashMap;

/// Renderer for evaluating Natsuzora AST
pub struct Renderer {
    template_loader: Option<TemplateLoader>,
}

impl Renderer {
    /// Create a new renderer
    pub fn new(template_loader: Option<TemplateLoader>) -> Self {
        Self { template_loader }
    }

    /// Render a template with the given data
    pub fn render(&mut self, template: &Template, data: Value) -> Result<String> {
        let mut context = Context::new(data)?;
        self.render_nodes(template.nodes(), &mut context)
    }

    fn render_nodes(&mut self, nodes: &[AstNode], context: &mut Context) -> Result<String> {
        let mut output = String::new();

        for node in nodes {
            match node {
                AstNode::Text(n) => output.push_str(&n.content),
                AstNode::Variable(n) => output.push_str(&self.render_variable(n, context)?),
                AstNode::Unsecure(n) => output.push_str(&self.render_unsecure(n, context)?),
                AstNode::If(n) => output.push_str(&self.render_if(n, context)?),
                AstNode::Unless(n) => output.push_str(&self.render_unless(n, context)?),
                AstNode::Each(n) => output.push_str(&self.render_each(n, context)?),
                AstNode::Include(n) => output.push_str(&self.render_include(n, context)?),
            }
        }

        Ok(output)
    }

    fn render_variable(&self, node: &VariableNode, context: &Context) -> Result<String> {
        let location = node.location;
        let value = context.resolve(node.path.segments(), location)?;

        let str_value = match node.modifier {
            Modifier::None => value.stringify()?,
            Modifier::Nullable => value.stringify_nullable()?,
            Modifier::Required => value.stringify_required()?,
        };
        Ok(html_escape::escape(&str_value))
    }

    fn render_unsecure(&self, node: &UnsecureNode, context: &Context) -> Result<String> {
        let location = node.location;
        let value = context.resolve(node.path.segments(), location)?;
        value.stringify()
    }

    fn render_if(&mut self, node: &IfBlock, context: &mut Context) -> Result<String> {
        let location = node.location;
        let value = context.resolve(node.condition.segments(), location)?;

        if value.is_truthy() {
            self.render_nodes(&node.then_branch, context)
        } else if let Some(else_branch) = &node.else_branch {
            self.render_nodes(else_branch, context)
        } else {
            Ok(String::new())
        }
    }

    fn render_unless(&mut self, node: &UnlessBlock, context: &mut Context) -> Result<String> {
        let location = node.location;
        let value = context.resolve(node.condition.segments(), location)?;

        if value.is_truthy() {
            Ok(String::new())
        } else {
            self.render_nodes(&node.body, context)
        }
    }

    fn render_each(&mut self, node: &EachBlock, context: &mut Context) -> Result<String> {
        let location = node.location;

        let items: Vec<Value> = match context.resolve(node.collection.segments(), location)? {
            Value::Array(arr) => arr.clone(),
            other => {
                return Err(NatsuzoraError::TypeError {
                    message: format!("Expected array, got {}", other.type_name()),
                });
            }
        };

        let mut output = String::new();
        for item in items {
            let mut bindings = HashMap::new();
            bindings.insert(node.item_ident.clone(), item);

            context.push_scope(bindings)?;
            let iteration = self.render_nodes(&node.body, context)?;
            context.pop_scope();

            output.push_str(&iteration);
        }

        Ok(output)
    }

    fn render_include(&mut self, node: &IncludeNode, context: &mut Context) -> Result<String> {
        let partial = self
            .template_loader
            .as_mut()
            .ok_or_else(|| NatsuzoraError::IncludeError {
                message: "Template loader not configured for include".to_string(),
            })?
            .load(&node.name)?;

        let mut bindings = HashMap::new();
        for arg in &node.args {
            let value = context.resolve(arg.value.segments(), arg.location)?.clone();
            bindings.insert(arg.name.clone(), value);
        }

        if let Some(loader) = self.template_loader.as_mut() {
            loader.push_include(&node.name);
        }

        context.push_include_scope(bindings);
        let result = self.render_nodes(partial.nodes(), context);
        context.pop_scope();

        let result = result.map_err(|e| match e {
            NatsuzoraError::WithIncludeTrace { .. } => e,
            e => match self.template_loader.as_ref() {
                Some(loader) => NatsuzoraError::WithIncludeTrace {
                    trace: loader.include_stack_trace(),
                    source: Box::new(e),
                },
                None => e,
            },
        });

        if let Some(loader) = self.template_loader.as_mut() {
            loader.pop_include();
        }

        result
    }
}
